### DISCUSSION TITLE
why optimization doesn't work

### DISCUSSION THREAD

**Discussion by EXing** [facebookresearch/theseus#603] [Q&A] [answered]
code from #596 
```
Start theseus iteration:
Nonlinear optimizer. Iteration: 0. Error: 8258.4384765625
Nonlinear optimizer. Iteration: 1. Error: 8258.4384765625
End theseus iteration.
```
why was the optimization finished directly?

**Comment by EXing** [comment]
how can I find out which end condition is triggered?

**Comment by mhmukadam** [comment]
You could setup breakpoints in [this function](https://github.com/facebookresearch/theseus/blob/main/theseus/optimizer/nonlinear/nonlinear_optimizer.py#L110) to see which condition is triggered. The error stays the same between the two iterations so likely it's [the first one](https://github.com/facebookresearch/theseus/blob/main/theseus/optimizer/nonlinear/nonlinear_optimizer.py#L111) (error abs change is less than threshold).

**Comment by luisenp** [comment, marked as answer]
@EXing The `TheseusLayer.forward()` method [returns](https://github.com/facebookresearch/theseus/blob/main/theseus/theseus_layer.py#L97) and info object that has status codes for termination conditions. When converging in one step a common cause is 0 gradients leading to immediate (incorrect) convergence.

**Comment by EXing** [comment]
```
# For now, step size is combined with delta. If we add more sophisticated
# line search, will probably need to pass it separately, or compute inside.
err, all_rejected = self._step(
    delta * steps_tensor,
    info.last_err,
    converged_indices,
    force_update,
    delta_forced_to_gn=_last_implicit_diff_step,
    **kwargs,
)  # err is shape (batch_size,)
```
all_rejected: True
converged_indices: [False]
force_update: False
_last_implicit_diff_step: False

Find out all update was rejected, any help?

**Comment by EXing** [comment]
`reject_indices = self._complete_step(delta, err, previous_err, **kwargs)`
```
    def _complete_step(
        self,
        delta: torch.Tensor,
        new_err: torch.Tensor,
        previous_err: torch.Tensor,
        adaptive_damping: bool = False,
        down_damping_ratio: float = 9.0,
        up_damping_ratio: float = 11.0,
        damping_accept: float = 0.1,
        ellipsoidal_damping: bool = False,
        **kwargs,
    ) -> Optional[torch.Tensor]:
        # "err" tensors passed as input refer to the squared norm of the
        # error vector, as returned by self.objective.error_metric()
        if adaptive_damping:
            return self._check_accept(
                delta,
                new_err,
                previous_err,
                damping_accept,
                down_damping_ratio,
                up_damping_ratio,
                ellipsoidal_damping,
            )
        else:
            return None
```
seems adaptive damping doesn't work

**Comment by EXing** [comment]
```
    optimizer = th.LevenbergMarquardt(
        problem.objective,
        linear_solver_cls=th.LUDenseSolver,
        linearization_cls=th.DenseLinearization,
        step_size=1e-2,
        max_iterations=50,
    )
    theseus_layer = th.TheseusLayer(optimizer)
    with torch.no_grad():
        with global_map.keyframes_mutex:
            print("Start theseus iteration:")
            theseus_outputs, info = theseus_layer.forward(optimizer_kwargs={
                "verbose": global_map.cfg.inner_optim.verbose,
                "adaptive_damping": True,
            })
            print("End theseus iteration.")
```
This works, but very slow with many iterations

**Reply by luisenp** [reply]
Hi @EXing do you have a short repro script so I can take a closer look?

**Reply by EXing** [reply]
Sorry, it's hard to reproduce with a short script.
The project contains multiple files and needs to download datasets.
I can email you a project zip file.
